### PR TITLE
Support fixed and jittered retry backoff policy

### PR DESCRIPTION
**Pull request by chenBright**
### What problem does this PR solve?

Issue Number:

Problem Summary:

支持固定/随机时间间隔重试退避策略：对于一些暂时性的错误，如网络抖动等，延迟一小会儿再重试的成功率会比立即重试的成功率高，同时可以打散上游重试的时机，减轻服务端压力，避免重试风暴导致服务端出现瞬间流量洪峰。

### What is changed and the side effects?

Changed:

Side effects:
- Performance effects(性能影响):

- Breaking backward compatibility(向后兼容性): 

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

File: src/brpc/controller.cpp [hunk 1]
```diff
@@ -1013,6 +1020,18 @@ void Controller::IssueRPC(int64_t start_realtime_us) {
         return;
     }
 
+    bthread::TaskGroup* g = bthread::tls_task_group;
+    // Only bthread task can retry with backoff.
```
Comment (wwbmmm) [inline src/brpc/controller.cpp:1024]: 如果检测到是pthread task，报个错？

File: src/brpc/retry_backoff_policy.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace brpc {
+
+// Inherit this class to customize the RPC retry backoff policy.
+class RetryBackoffPolicy {
+public:
+    virtual ~RetryBackoffPolicy() = default;
+
+    // Returns the backoff time in milliseconds before every retry.
+    virtual int32_t get_backoff_time_ms(int nretry, int64_t reaming_rpc_time_ms) const = 0;
```
[the remark this answers is not in the document]
Comment (chenBright) [reply src/brpc/retry_backoff_policy.h]: done

File: src/brpc/retry_backoff_policy.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    //                                                              don't forget the const modifier
+};
+
+class FixedRetryBackoffPolicy : public RetryBackoffPolicy {
+public:
+    FixedRetryBackoffPolicy(int32_t backoff_time_ms, int32_t no_backoff_remaining_rpc_time_ms)
+        : _backoff_time_ms(backoff_time_ms)
+        , _no_backoff_remaining_rpc_time_ms(no_backoff_remaining_rpc_time_ms) {}
+
+    int32_t get_backoff_time_ms(int nretry, int64_t reaming_rpc_time_ms) const override;
```
[the remark this answers is not in the document]
Comment (chenBright) [reply src/brpc/retry_backoff_policy.h]: done

File: test/brpc_channel_unittest.cpp [hunk 4]
```diff
@@ -1810,6 +1811,73 @@ class ChannelTest : public ::testing::Test{
         StopAndJoin();
     }
 
+    struct TetsRetryBackoff {
```
[the remark this answers is not in the document]
Comment (chenBright) [reply test/brpc_channel_unittest.cpp]: done

File: src/brpc/controller.cpp [hunk 5]
```diff
@@ -1013,6 +1020,18 @@ void Controller::IssueRPC(int64_t start_realtime_us) {
...
+    if (g && !g->is_current_pthread_task() && _retry_backoff && _current_call.nretry > 0) {
+        int64_t remaining_rpc_time_us = _deadline_us - butil::gettimeofday_us();
```
[the remark this answers is not in the document]
Comment (chenBright) [reply src/brpc/controller.cpp]: done

File: src/brpc/controller.cpp (hunk 1 above)
Comment (chenBright) [reply src/brpc/controller.cpp]: ChannelOptions增加一个开关，支持在pthread进行重试退避。
没打开开关，但是需要在pthread task中重试退避时，跳过且打印警告日志。

File: src/brpc/channel.h [hunk 6]
```diff
@@ -119,6 +120,17 @@ struct ChannelOptions {
...
+    // Backoff request before every retry. The interface
+    // is defined in src/brpc/retry_backoff.h
+    // This object is NOT owned by channel and should
+    // remain valid when channel is used.
+    // Default: NULL
+    const RetryBackoffPolicy* retry_backoff_policy;
+    // Enable retry backoff in pthread.
+    // Note it will block the pthread.
+    // Default: false
+    bool enable_retry_backoff_in_pthread;
```
Comment (wwbmmm) [inline src/brpc/channel.h:132]: 这个参数，是否可以给RetryBackoffPolicy加一个虚方法来返回？
这样就不用加太多的参数在ChannelOptions里

**Comment by wwbmmm** [comment]
@chenBright 麻烦再补充下使用文档说明吧

**Comment by thorneliu** [comment]
Hi,

Since `Retry_backoff_policy` or 'Backoff_retry_policy' **IS A** Retry policy, it is not a good idea 
that both policy exist in channel, 
```
    , retry_policy(NULL)
    , retry_backoff_policy(NULL)
```

It is better that retry_backoff_policy inherits basic interface in a manner of :

```
class BackoffRetryPolicy : public RetryPolicy {
 bool DoRetry(const Controller* controller) const {
   // details in backoff policy
 }
}
```

**Comment by lorinlee** [comment]
或者是不是也可以在RetryPolicy里加一些接口

**Comment by chenBright** [comment]
> Hi,
> 
> Since `Retry_backoff_policy` or `Backoff_retry_policy` **IS A** Retry policy, it is not a good idea that both policies exist in channel,
> 
> ```
>     , retry_policy(NULL)
>     , retry_backoff_policy(NULL)
> ```
> 
> It is better that retry_backoff_policy inherits basic interface in a manner of :
> 
> ```
> class BackoffRetryPolicy : public RetryPolicy {
>  bool DoRetry(const Controller* controller) const {
>    // details in backoff policy
>  }
> }
> ```

@thorneliu Good idea. Maybe These virtual functions can be added to RetryPolicy with default implementation that the retry backoff feature is turned off.  In a manner of:

``` c++
class RetryPolicy {
public:
    virtual ~RetryPolicy();
    virtual bool DoRetry(const Controller* controller) const = 0;
    virtual int32_t GetBackoffTimeMs(const Controller* controller, int nretry,
                                     int64_t remaining_rpc_time_ms) const { return 0; }
    virtual bool CanRetryBackoffInPthread() const { return false; }
};
```

`FixedRetryBackoffPolicy` and `JitteredRetryBackoffPolicy` inherit `RpcRetryPolicy` which implements the default rpc retry policy.

**Comment by chenBright** [comment]
> 或者是不是也可以在RetryPolicy里加一些接口

可以的，这样应该合理一些。

**Comment by chenBright** [comment]
> @chenBright 麻烦再补充下使用文档说明吧

@wwbmmm 按照 @thorneliu 意见修改了实现，麻烦再看看。没问题，我再补充使用文档。

**Comment by chenBright** [comment]
> > Hi,
> > Since `Retry_backoff_policy` or `Backoff_retry_policy` **IS A** Retry policy, it is not a good idea that both policies exist in channel,
> > ```
> >     , retry_policy(NULL)
> >     , retry_backoff_policy(NULL)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It is better that retry_backoff_policy inherits basic interface in a manner of :
> > ```
> > class BackoffRetryPolicy : public RetryPolicy {
> >  bool DoRetry(const Controller* controller) const {
> >    // details in backoff policy
> >  }
> > }
> > ```
> 
> @thorneliu Good idea. Maybe These virtual functions can be added to RetryPolicy with default implementation that the retry backoff feature is turned off. In a manner of:
> 
> ```c++
> class RetryPolicy {
> public:
>     virtual ~RetryPolicy();
>     virtual bool DoRetry(const Controller* controller) const = 0;
>     virtual int32_t GetBackoffTimeMs(const Controller* controller, int nretry,
>                                      int64_t remaining_rpc_time_ms) const { return 0; }
>     virtual bool CanRetryBackoffInPthread() const { return false; }
> };
> ```
> 
> `FixedRetryBackoffPolicy` and `JitteredRetryBackoffPolicy` inherit `RpcRetryPolicy` which implements the default rpc retry policy.

@thorneliu New implementationhas been updated. Could you spare some tine to review it again and offer any additional input?

**Comment by thorneliu** [comment]
> > > Hi,
> > > Since `Retry_backoff_policy` or `Backoff_retry_policy` **IS A** Retry policy, it is not a good idea that both policies exist in channel,
> > > ```
> > >     , retry_policy(NULL)
> > >     , retry_backoff_policy(NULL)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It is better that retry_backoff_policy inherits basic interface in a manner of :
> > > ```
> > > class BackoffRetryPolicy : public RetryPolicy {
> > >  bool DoRetry(const Controller* controller) const {
> > >    // details in backoff policy
> > >  }
> > > }
> > > ```
> > 
> > 
> > @thorneliu Good idea. Maybe These virtual functions can be added to RetryPolicy with default implementation that the retry backoff feature is turned off. In a manner of:
> > ```c++
> > class RetryPolicy {
> > public:
> >     virtual ~RetryPolicy();
> >     virtual bool DoRetry(const Controller* controller) const = 0;
> >     virtual int32_t GetBackoffTimeMs(const Controller* controller, int nretry,
> >                                      int64_t remaining_rpc_time_ms) const { return 0; }
> >     virtual bool CanRetryBackoffInPthread() const { return false; }
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `FixedRetryBackoffPolicy` and `JitteredRetryBackoffPolicy` inherit `RpcRetryPolicy` which implements the default rpc retry policy.
> 
> @thorneliu New implementationhas been updated. Could you spare some tine to review it again and offer any additional input?

Hi, 不好意思没有及时回复

我觉得最好是RetryPolicy的接口不添加 GetBackoffTimeMs 或CanRetryBackoffInPthread

我们是否可以做到，
```
fooPolicy = new brpc::fooPolicy(arg1, arg2).
channeloptions.retryPolicy = fooPolicy;
````
然后类似 BackoffTimeMs 这样的细节封装在各自的policy内部。基本的rpcRetryPolicy接口只暴露一个简单的 `DoRetry()` 方法？

**Comment by chenBright** [comment]
> > > > Hi,
> > > > Since `Retry_backoff_policy` or `Backoff_retry_policy` **IS A** Retry policy, it is not a good idea that both policies exist in channel,
> > > > ```
> > > >     , retry_policy(NULL)
> > > >     , retry_backoff_policy(NULL)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > It is better that retry_backoff_policy inherits basic interface in a manner of :
> > > > ```
> > > > class BackoffRetryPolicy : public RetryPolicy {
> > > >  bool DoRetry(const Controller* controller) const {
> > > >    // details in backoff policy
> > > >  }
> > > > }
> > > > ```
> > > 
> > > 
> > > @thorneliu Good idea. Maybe These virtual functions can be added to RetryPolicy with default implementation that the retry backoff feature is turned off. In a manner of:
> > > ```c++
> > > class RetryPolicy {
> > > public:
> > >     virtual ~RetryPolicy();
> > >     virtual bool DoRetry(const Controller* controller) const = 0;
> > >     virtual int32_t GetBackoffTimeMs(const Controller* controller, int nretry,
> > >                                      int64_t remaining_rpc_time_ms) const { return 0; }
> > >     virtual bool CanRetryBackoffInPthread() const { return false; }
> > > };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > `FixedRetryBackoffPolicy` and `JitteredRetryBackoffPolicy` inherit `RpcRetryPolicy` which implements the default rpc retry policy.
> > 
> > 
> > @thorneliu New implementationhas been updated. Could you spare some tine to review it again and offer any additional input?
> 
> Hi, 不好意思没有及时回复
> 
> 我觉得最好是RetryPolicy的接口不添加 GetBackoffTimeMs 或CanRetryBackoffInPthread
> 
> 我们是否可以做到，
> 
> ```
> fooPolicy = new brpc::fooPolicy(arg1, arg2).
> channeloptions.retryPolicy = fooPolicy;
> ```
> 
> 然后类似 BackoffTimeMs 这样的细节封装在各自的policy内部。基本的rpcRetryPolicy接口只暴露一个简单的 `DoRetry()` 方法？

https://github.com/apache/brpc/blob/f27fbb4989e1bb2542596049f0263f234b8afb17/src/brpc/controller.cpp#L630-L655

目前来看，好像不能只通过`DoRetry`实现backoff，框架中`DoRetry`的语义是是否要重试，如果是，还有一些工作要做：
``` c++
_current_call.OnComplete(this, _error_code, info.responded, false);
++_current_call.nretry;
```
如果在`DoRetry`中做backoff，这些工作也会被推迟，不太合适，还是做完这些工作再做backoff合适一些吧。而且用户要实现自定义backoff的话，就得需要十分清楚框架此处的实现（例如pthread上做backoff会阻塞pthread、_current_call.nretry还未加1、之前的rpc出现网络错误，socket被SetFailed的时机也推迟等），否则很容易会踩坑。

File: src/brpc/controller.cpp [hunk 7]
```diff
@@ -625,33 +630,49 @@ void Controller::OnVersionedRPCReturned(const CompletionInfo& info,
...
-        if (!SingleServer()) {
-            if (_accessed == NULL) {
-                _accessed = ExcludedServers::Create(
-                    std::min(_max_retry, RETRY_AVOIDANCE));
-                if (NULL == _accessed) {
-                    SetFailed(ENOMEM, "Fail to create ExcludedServers");
-                    goto END_OF_RPC;
+    } else {
+        auto retry_policy = _retry_policy ? _retry_policy : DefaultRetryPolicy();
+        if (retry_policy && retry_policy->DoRetry(this)) {
```
Comment (wwbmmm) [inline src/brpc/controller.cpp:635]: 按原来的代码逻辑，无须判断if (retry_policy)

**Comment by wwbmmm** [comment]
Macos编译失败，好像是protobuf的版本发生了变化？

**Comment by chenBright** [comment]
> Macos编译失败，好像是protobuf的版本发生了变化？

可能是

**Comment by wwbmmm** [comment]
master已修复 macos编译失败，可更新 @chenBright

File: src/brpc/controller.cpp (hunk 7 above)
Comment (chenBright) [reply src/brpc/controller.cpp]: done

**Comment by chenBright** [comment]
@wwbmmm 已更新使用文档

**Comment by wwbmmm** [comment]
LGTM

File: src/brpc/controller.cpp [hunk 8]
```diff
@@ -627,33 +632,51 @@ void Controller::OnVersionedRPCReturned(const CompletionInfo& info,
...
+            response_attachment().clear();
+
+            // Retry backoff.
+            bthread::TaskGroup* g = bthread::tls_task_group;
+            if (retry_policy->CanRetryBackoffInPthread() ||
+                (g && !g->is_current_pthread_task())) {
+                int64_t backoff_time_us = retry_policy->GetBackoffTimeMs(this) * 1000L;
+                // No need to do retry backoff when the backoff time is longer than the remaining rpc time.
+                if (backoff_time_us > 0 &&
+                    backoff_time_us < _deadline_us - butil::gettimeofday_us()) {
```
[the remark this answers is not in the document]
Comment (chenBright) [reply src/brpc/controller.cpp:670]: 嗯，统一在controller里判断吧，这是框架的约束，由框架来保证。

另外删除Retry里对重试次数的判断，框架保证了重试时才会调GetBackoffTimeMs。

File: src/brpc/retry_policy.cpp [hunk 9]
```diff
@@ -58,4 +54,28 @@ const RetryPolicy* DefaultRetryPolicy() {
...
+        return 0;
+    }
+    return _backoff_time_ms < remaining_rpc_time_ms ? _backoff_time_ms : 0;
+}
+
+int32_t RpcRetryPolicyWithJitteredBackoff::GetBackoffTimeMs(const Controller* controller) const {
+    if (controller->retried_count() <= 0) {
+        return 0;
+    }
+    int64_t remaining_rpc_time_ms = controller->deadline_us() / 1000 - butil::gettimeofday_ms();
```
[the remark this answers is not in the document]
Comment (chenBright) [reply src/brpc/retry_policy.cpp]: done

**Comment by serverglen** [comment]
LGTM

**Pull request merged**